File: GildedRoseKata/scala/src/test/scala/com/gildedrose/BackstagePassesSpec.scala
package com.gildedrose

import org.scalacheck.Properties
import org.scalacheck.Prop.forAll
import Generators._

class BackstagePassesSpec extends Properties("BackstagePasses"){

  property("Backstage Passes increase in quality by one if expiration is > 10 days away") = forAll(backstagePassesFarGen) {
    case (days: Int, backstagePassesFar: Item) => {
      (0 until days).map((_: Int) => {
          val beforeUpdateQuality = backstagePassesFar.quality
          val result = GildedRose.updateQuality(Array(backstagePassesFar))
          (result.length == 1 && (result(0).quality == (beforeUpdateQuality + 1) ||
            (result(0).quality == 50 && beforeUpdateQuality == 50)))
        }).forall(_ == true)
    }
  }

  property("Backstage Passes increase in quality by two if expiration is between 5-10 days away") = forAll(backstagePassesCloseGen) {
    case (days: Int, backstagePassesClose: Item) => {
      (0 until days).map((d: Int) => {
          val beforeUpdateQuality = backstagePassesClose.quality
          val result = GildedRose.updateQuality(Array(backstagePassesClose))
          (result.length == 1 && (result(0).quality == (beforeUpdateQuality + 2) ||
            (result(0).quality == 50 && beforeUpdateQuality == 49) ||
            (result(0).quality == 50 && beforeUpdateQuality == 50)))
        }).forall(_ == true)
    }
  }

  property("Backstage Passes increase in quality by three if expiration is between 0-5 days away") = forAll(backstagePassesClosestGen) {
    case (days: Int, backstagePassesClosest: Item) => {
      (0 until days).map((d: Int) => {
          val beforeUpdateQuality = backstagePassesClosest.quality
          val result = GildedRose.updateQuality(Array(backstagePassesClosest))
          (result.length == 1 && (result(0).quality == (beforeUpdateQuality + 3) ||
            (result(0).quality == 50 && beforeUpdateQuality == 48) ||
            (result(0).quality == 50 && beforeUpdateQuality == 49) ||
            (result(0).quality == 50 && beforeUpdateQuality == 50)))
        }).forall(_ == true)
    }
  }

  property("Backstage Passes drop quality to 0 if expiration happens") = forAll(backstagePassesWillExireGen) {
    case (days: Int, backstagePassesWillExpire: Item) => {
      (0 until days).map((d: Int) => {
          val result = GildedRose.updateQuality(Array(backstagePassesWillExpire))
          (result.length == 1 && result(0).quality == 0)
        }).dropWhile(_ == false).forall(_ == true)
    }
  }
}

File: GildedRoseKata/haskell/test/BackstagePassesSpec.hs
module BackstagePassesSpec (spec) where

import GildedRose
import Item
import Generators
import Test.Hspec
import Test.QuickCheck
import SuccessConditions
import PropertyChecks

spec :: Spec
spec = describe "BackstagePassesSpec" $ do
  describe "should increase in quality" $ do
    it "by one if the expiration date is far away (> 10)" $ property $
      forAll backstagePassesFarGen $ qualityCheckSingleton ascendingQualty1
    it "by two if the expiration date is close (5-10)" $ property $
      forAll backstagePassesCloseGen $ qualityCheckSingleton ascendingQualty2
    it "by three if the expiration date is closest (0-5)" $ property $
      forAll backstagePassesCloseGen $ qualityCheckSingleton ascendingQualty3
  describe "should have quality == 0" $
    it "When it expires" $
      forAll backstagePassesWillExpireGen
        (\(i, d)->
          let days = getPositive d
              successCondition r = length r == 1 && 0 == (valueQ . getQuality . head) r
              result = foldl (\acc iteration -> acc || successCondition iteration) False $ take days (iterate updateQuality [i])
          in result)
